### PR TITLE
U5 - ASR

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
Public repo version of Unit 5, supersedes the private repo version #26

### CODE REVIEW

File: chapters/en/chapter5/introduction.mdx [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+You’ve probably made use of a speech recognition system many times before without realising! Consider the digital
+assistant in your smartphone device (Siri, Google Assistant, Alexa). When you use these assistants, the first thing that
+they do is transcribe your spoken speech to written text, ready to be used for any downstream tasks (such as finding you
+the weather 🌤️).
+
+Have a play with the speech recognition demo below. You can either record yourself using your microphone, or drag and
+drop an audio sample for transcription:
+
+<gradio-app space="sanchit-gandhi/whisper-small"></gradio-app>
```
Comment (MKhalusova) [inline chapters/en/chapter5/introduction.mdx:21]: Please change to: 
<iframe
	src="https://sanchit-gandhi-whisper-small.hf.space"
	frameborder="0"
	width="850"
	height="450">
</iframe>

File: chapters/en/chapter5/evaluation.mdx [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+
+Here, we have:
+* 1 substitution ("sit" instead of "sat")
+* 0 insertions
+* 1 deletion ("mat" is missing)
+
+This gives 2 errors in total. To get our error rate, we divide the number of errors by the total number of words in our
+reference (N), which for this example is 6:
+
+\begin{aligned}
```
Comment (MKhalusova) [inline chapters/en/chapter5/evaluation.mdx:49]: The formula doesn't render.

File: chapters/en/chapter5/evaluation.mdx [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+|-------------|-----|-----|-----|-----|-----|-----|-----|-----|-----|-------|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|
+| Prediction: | t   | h   | e   |     | c   | a   | t   |     | s   | **i** | t   |     | o   | n   |     | t   | h   | e   |     |     |     |     |
+| Label:      | ✅   | ✅   | ✅   |     | ✅   | ✅   | ✅   |     | ✅   | S     | ✅   |     | ✅   | ✅   |     | ✅   | ✅   | ✅   |     | D   | D   | D   |
+
+We can see now that for the word "sit", the "s" and "t" are marked as correct! It's only the "i" which is labelled as a
+substitution error (S). Thus, we reward our system for the partially correct prediction 🤝
+
+In our example, we have 1 character substitution, 0 insertions, and 3 deletions. In total, we have 14 characters. So, our CER is:
+
+\begin{aligned}
```
Comment (MKhalusova) [inline chapters/en/chapter5/evaluation.mdx:121]: Same as earlier: the formula doesn't render

File: chapters/en/chapter5/evaluation.mdx [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+```python
+cer_metric = load("cer")
+
+cer = cer_metric.compute(references=[reference], predictions=[prediction])
+
+print(cer)
+```
+**Print Output:**
+```
+TODO(SG)
```
Comment (MKhalusova) [inline chapters/en/chapter5/evaluation.mdx:140]: A leftover TODO

File: chapters/en/chapter5/evaluation.mdx [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+</Tip>
+
+And finally, we can compute the WER. Let's first compute the orthographic WER, i.e. the WER without any post-processing:
+
+```python
+from evaluate import load
+
+wer_metric = load("wer")
+
+wer_ortho = 100 * wer_metric.compute(
```
Comment (MKhalusova) [inline chapters/en/chapter5/evaluation.mdx:334]: Why do we switch to percentages? (multiplying by 100)

File: chapters/en/chapter5/choosing_dataset.mdx [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+Dhivehi for this tutorial, the steps covered here apply to any one of the 108 languages in the Common Voice 13 dataset, and
+more generally to any one of the 180+ audio datasets on the Hugging Face Hub, so there's no restriction on language or dialect.
+
+We can select the Dhivehi subset of Common Voice 13 by setting the subset to `dv` using the dropdown menu (`dv` being the language
+identifier code for Dhivehi):
+
+<!--- Fast render, don't display figs
+
+<figure>
+<img src=TODO alt="Trulli" style="width:100%">
```
Comment (MKhalusova) [inline chapters/en/chapter5/choosing_dataset.mdx:111]: An image missing

File: chapters/en/chapter5/fine-tuning.mdx [hunk 7]
```diff
@@ -0,0 +1,549 @@
...
+your GPU, it is possible that you will encounter a CUDA `"out-of-memory"` error when you start training. In this case,
+you can reduce the `per_device_train_batch_size` incrementally by factors of 2 and employ [`gradient_accumulation_steps`](https://huggingface.co/docs/transformers/main_classes/trainer#transformers.Seq2SeqTrainingArguments.gradient_accumulation_steps)
+to compensate.
+
+**Output:**
+
+| Step | Training Loss | Epoch | Validation Loss |  WER  |
+|:----:|:-------------:|:-----:|:---------------:|:-----:|
+
+Our best WER is X% TODO(SG) - not bad for 7h of training data (I hope...)! The big question is how this
```
Comment (MKhalusova) [inline chapters/en/chapter5/fine-tuning.mdx:492]: Missing WER

File: chapters/en/_toctree.yml [hunk 8]
```diff
@@ -65,29 +65,29 @@
...
 #  - local: chapter5/speaker_diarization
 #    title: Automatic speech recognition with speaker diarization
-#  - local: chapter5/quiz
-#    title: Quiz
-#    quiz: 5
-#  - local: chapter5/hands_on
-#    title: Hands-on exercise
-#  - local: chapter5/supplemental_reading
-#    title: Supplemental reading and resources
+  - local: chapter4/demo
```
Comment (MKhalusova) [inline chapters/en/_toctree.yml:82]: Please remove the duplicate of the page from the TOC (building a demo)

File: chapters/en/chapter5/introduction.mdx (hunk 1 above)
Comment (sanchit-gandhi) [reply chapters/en/chapter5/introduction.mdx:21]:
```suggestion
<iframe src="https://sanchit-gandhi-whisper-small.hf.space" frameborder="0" width="850" height="450"> </iframe>
```

File: chapters/en/chapter5/evaluation.mdx (hunk 5 above)
Comment (sanchit-gandhi) [reply chapters/en/chapter5/evaluation.mdx:334]: For consistency with our previous scripts (e.g. the Whisper event) where we returned the WER in % (see https://huggingface.co/spaces/whisper-event/leaderboard?dataset=mozilla-foundation%2Fcommon_voice_11_0&config=ar&split=test)

Agnostic as to whether we do it in absolute or percentage, but the latter is more consistent with the scripts we've published in the past

File: chapters/en/chapter5/evaluation.mdx (hunk 2 above)
Comment (sanchit-gandhi) [reply chapters/en/chapter5/evaluation.mdx:50]: Resolved in https://github.com/huggingface/audio-transformers-course/pull/70/commits/2d0ab9194c416f9ea4b0e78e68311e9b02332e65

File: chapters/en/chapter5/evaluation.mdx (hunk 3 above)
Comment (sanchit-gandhi) [reply chapters/en/chapter5/evaluation.mdx:127]: Resolved in https://github.com/huggingface/audio-transformers-course/pull/70/commits/2d0ab9194c416f9ea4b0e78e68311e9b02332e65

File: chapters/en/chapter5/evaluation.mdx (hunk 4 above)
Comment (sanchit-gandhi) [reply chapters/en/chapter5/evaluation.mdx]: Resolved in https://github.com/huggingface/audio-transformers-course/pull/70/commits/f0e626f03e6e9c3625f5335d67040bc6e1a0c858

File: chapters/en/chapter5/choosing_dataset.mdx (hunk 6 above)
Comment (sanchit-gandhi) [reply chapters/en/chapter5/choosing_dataset.mdx]: Resolved in https://github.com/huggingface/audio-transformers-course/pull/70/commits/b922ae0862a6c6394ec7b2b52966f13fdf9321ea

File: chapters/en/chapter5/fine-tuning.mdx [hunk 9]
```diff
@@ -0,0 +1,549 @@
...
+compares to other ASR systems. For that, we can view the autoevaluate [leaderboard](https://huggingface.co/spaces/autoevaluate/leaderboards?dataset=mozilla-foundation%2Fcommon_voice_13_0&only_verified=0&task=automatic-speech-recognition&config=hi&split=test&metric=wer),
+a leaderboard that categorises models by language and dataset, and subsequently ranks them according to their WER.
+
+We see...
```
[the remark this answers is not in the document]
Comment (sanchit-gandhi) [reply chapters/en/chapter5/fine-tuning.mdx]: Resolved in https://github.com/huggingface/audio-transformers-course/pull/70/commits/9171b6c80f0b4f0fbd382eb75cf2336a919683b6

File: chapters/en/_toctree.yml (hunk 8 above)
Comment (sanchit-gandhi) [reply chapters/en/_toctree.yml]: Ahh good catch! Added a new section on building a demo for U5 https://github.com/huggingface/audio-transformers-course/pull/70/commits/cec9a630636fe8f58c445d8acb1d2fcedf9ffe73

File: chapters/en/chapter5/fine-tuning.mdx (hunk 7 above)
Comment (sanchit-gandhi) [reply chapters/en/chapter5/fine-tuning.mdx]: Resolved in https://github.com/huggingface/audio-transformers-course/pull/70/commits/9171b6c80f0b4f0fbd382eb75cf2336a919683b6

**Pull request merged**